### PR TITLE
Implement Maniacs Command 3009: ControlBattle

### PR DESCRIPTION
**Pull request by MakoInfused**
With all of these Control Battle events set:
image
image
image
image
image
image

Game developers have a way to inject events hooks into the battle system that might look something like this:
EasyRPG-ManiacsControlBattleCommand

This allows them to handle all kinds of events including:
- ATB Changes (parallel process like event)
- Health Changes (custom damage popup visualization)
- Targets being selected (before turn event)
- State Changes (custom status effects)
- Other Stat Changes (custom stat change popup visualization)

One small difference from the maniacs version (Maniacs 210414) I use:
🐛  I unintentionally fixed a bug: when the battle begins in Maniacs, it only fire an event for the first state to change (or maybe it's just not waiting for the message box to close for the others?).
✅  In EasyRPG it will definitely provide the callback for all state changes, regardless of if it's at the beginning of the battle or during it and regardless of how many state changes happened. However, it works the same as maniacs in that it doesn't block the battle from ending in order to complete a callback.

Everything else should match identically. 

My Testing Included:
1. Every single variable value is sets before the callbacks.
2. Every single callback.
3. Every single callbacks handling of things that block it, such as event messages.

There was a lot to test with many different permutation combinations, so it's always possible something slipped. If so, please let me know. :)

### CODE REVIEW

**Review by Ghabry — CHANGES_REQUESTED** [review]
What I havn't fully reviewed yet is this sub-interpreter stuff. Have to test this while running a real game in Maniacs vs Easy.

File: src/game_interpreter_battle.h [hunk 1]
```diff
@@ -60,6 +61,26 @@ class Game_Interpreter_Battle : public Game_Interpreter
 
 	bool ExecuteCommand(lcf::rpg::EventCommand const& com) override;
 
+	/**
+	 * All possible hook type events that maniacs offers.
+	 */
+	enum ManiacBattleHookType {
```
[the remark this answers is not in the document]
Comment (MakoInfused) [reply src/game_interpreter_battle.h:67]: Done

File: src/game_battlealgorithm.cpp [hunk 2]
```diff
@@ -849,6 +912,14 @@ Game_BattleAlgorithm::Skill::Skill(Game_Battler* source, const lcf::rpg::Skill&
 {
 }
 
+int Game_BattleAlgorithm::Skill::GetActionType() {
+	return 1;
```
[the remark this answers is not in the document]
Comment (MakoInfused) [reply src/game_battlealgorithm.cpp:926]: Okay, I've made the change as requested. 

Note that: technically these aren't exclusively enemy actions, they are for any battler, so in the future we could either: 

1. Rename the enum from "EnemyAction" -> "BattlerAction", then we can have a item 3. 
2. Create a new enum which is specifically for "ActorAction" and have that one include the same values as Enemy and adding the item 3.

I'll leave it for now and assume we'll do nothing. However, if you have a preference, let me know and I can easily make the change as you choose.

File: src/game_battle.h [hunk 3]
```diff
@@ -109,6 +110,15 @@ namespace Game_Battle {
 	 */
 	void UpdateAtbGauges();
 
+	/**
+	 * Convenience function to call a maniacs battle hook, which processes sub-events at any time.
+	 */
+	bool ManiacBattleHook(Game_Interpreter_Battle::ManiacBattleHookType hook_type, int var1, int var2, int var3, int var4 = 0, int var5 = 0, int var6 = 0);
```
[the remark this answers is not in the document]
Comment (MakoInfused) [reply src/game_battle.h:116]: I considered that, but thinking ahead: I'm not sure if BingShan might in the future introduce a new type which doesn't follow the pattern, so I decided to leave it generic.

File: src/scene_battle_rpg2k3.cpp [hunk 4]
```diff
@@ -498,7 +498,24 @@ void Scene_Battle_Rpg2k3::UpdateAnimations() {
...
+	ss >> value;
+	bool should_override = Game_Battle::ManiacBattleHook(
+		Game_Interpreter_Battle::DamagePop,
+		battler->GetType() == Game_Battler::Type_Enemy,
+		battler->GetPartyIndex(),
+		x,
+		y,
+		static_cast<int>(type),
+		value
+	);
```
[the remark this answers is not in the document]
Comment (MakoInfused) [reply src/scene_battle_rpg2k3.cpp:513]: Yes, it's tested and the value is always 0 when the 5th var (type of value) = 2.

File: src/game_interpreter_battle.cpp [hunk 5]
```diff
@@ -43,14 +43,35 @@ enum TargetType {
...
 
+// Implemented as a static map, since maniac hooks can only have one common event callback at a time.
+// Subsequent calls will simply override the previous common event callback.
+std::map<Game_Interpreter_Battle::ManiacBattleHookType, std::tuple<int, int>> Game_Interpreter_Battle::maniac_hooks = {
+		{AtbIncrement, std::make_tuple(0, 0)},
+		{DamagePop, std::make_tuple(0, 0)},
+		{Targetting, std::make_tuple(0, 0)},
+		{SetState, std::make_tuple(0, 0)},
+		{StatChange, std::make_tuple(0, 0)}
+};
```
[the remark this answers is not in the document]
Comment (MakoInfused) [reply src/game_interpreter_battle.cpp:54]: They are not preserved across battles, good catch, thank you! 

I've added a `Game_Interpreter_Battle::InitBattle` method which gets called from the battle scene during it's initialization. I've tested this and it correctly now resets on the start of each battle. 

I considered moving it into the current instance of the game battle interpreter, but I think that makes it seem like can be called for reach instance, when in reality it shouldn't (and now that I have sub interpreters, this is possible). I also considered moving it into the Game_Battle itself, but I don't really want to expose the hooks for modification from outside the interpreter either. 

Let me know if you have any questions.

**Comment by MakoInfused** [comment]
> What I havn't fully reviewed yet is this sub-interpreter stuff. Have to test this while running a real game in Maniacs vs Easy.

Sounds good, let me know if you need anything from me. I have a test project that I'm using if you like I can provide it for your testing. It's the same one I've used to create the gif in this pull request. I believe also @jetrotal tested it on Beloved Rapture, so using an existing project is also an option.

File: src/game_battlealgorithm.h [hunk 6]
```diff
@@ -83,6 +83,19 @@ class AlgorithmBase {
 public:
 	virtual ~AlgorithmBase() {}
 
+	enum ActionType {
+		Basic,
+		Skill,
+		Transformation,
+		Item
+	};
```
[the remark this answers is not in the document]
Comment (Ghabry) [reply src/game_battlealgorithm.h:91]: I think this enum is still unused? (and you can delete it)

**Comment by Ghabry** [comment]
besides that one comment this looks fine now. So after this no further code changes required by you. Thanks.

I will report back when I did some testing.

**Review by fdelapena — APPROVED** [review]

File: src/scene_battle_rpg2k3.cpp [hunk 7]
```diff
@@ -60,10 +60,13 @@ Scene_Battle_Rpg2k3::Scene_Battle_Rpg2k3(const BattleArgs& args) :
 
 void Scene_Battle_Rpg2k3::Start() {
 	Scene_Battle::Start();
+	Game_Interpreter_Battle::InitBattle();
 	InitBattleCondition(Game_Battle::GetBattleCondition());
 	CreateEnemySprites();
 	CreateActorSprites();
 
+	Game_Battle::GetInterpreterBattle();
```
Comment (Ghabry) [inline src/scene_battle_rpg2k3.cpp:68]: This function has no side effects and the return value is ignored. Can be removed?

**Comment by Ghabry** [comment]
Besides Maniacs being strange when a message box is open this works for me :+1: 

----

And I made one more change:

I didn't want to push this directly to your branch as I want to have 2 more eyes to look at the code if this breaks anything xD:

By moving the ``ManiacBattleHookType::Targetting`` to ``Scene_Battle::ActionSelectedCallback`` it will work for the 2000 battle system.

And then I copy-pasted the event processing to the 2k battle event loop to make the hooks execute in 2k.

Made a quick test and this makes all callbacks work in a 2k battle except damage pop (2k has no floating numbers) and ATB change (makes no sense for 2k).

```diff
diff --git a/src/scene_battle.cpp b/src/scene_battle.cpp
index aaf6219f0..d388e8f5a 100644
--- a/src/scene_battle.cpp
+++ b/src/scene_battle.cpp
@@ -570,6 +570,21 @@ void Scene_Battle::RemoveCurrentAction() {
 void Scene_Battle::ActionSelectedCallback(Game_Battler* for_battler) {
 	assert(for_battler->GetBattleAlgorithm() != nullptr);
 
+	auto single_target = for_battler->GetBattleAlgorithm()->GetOriginalSingleTarget();
+	auto group_targets = for_battler->GetBattleAlgorithm()->GetOriginalPartyTarget();
+	// Target: 0 None, 1 Single Enemy, 2 All Enemies, 3 Single Ally, 4 All Allies
+	Game_Battle::ManiacBattleHook(
+		Game_Interpreter_Battle::ManiacBattleHookType::Targetting,
+		for_battler->GetType() == Game_Battler::Type_Enemy,
+		for_battler->GetPartyIndex(),
+		for_battler->GetBattleAlgorithm()->GetActionType(),
+		for_battler->GetBattleAlgorithm()->GetActionId(),
+		single_target
+			? (single_target->GetType() != Game_Battler::Type_Enemy ? 1 : 3)
+			: (group_targets->GetRandomActiveBattler()->GetType() != Game_Battler::Type_Enemy ? 2 : 4),
+		single_target ? single_target->GetPartyIndex() : 0
+	);
+
 	if (for_battler->GetBattleAlgorithm() == nullptr) {
 		Output::Warning("ActionSelectedCallback: Invalid action for battler {} ({})",
 				for_battler->GetId(), for_battler->GetName());
diff --git a/src/scene_battle_rpg2k.cpp b/src/scene_battle_rpg2k.cpp
index 1026e3c9d..9d61a317d 100644
--- a/src/scene_battle_rpg2k.cpp
+++ b/src/scene_battle_rpg2k.cpp
@@ -205,6 +205,12 @@ void Scene_Battle_Rpg2k::vUpdate() {
 			break;
 		}
 
+		// this is checked separately because we want normal events to be processed
+		// just not sub-events called by maniacs battle hooks.
+		if (state != State_Victory && state != State_Defeat && Game_Battle::ManiacProcessSubEvents()) {
+			break;
+		}
+
 		if (!CheckWait()) {
 			break;
 		}
diff --git a/src/scene_battle_rpg2k3.cpp b/src/scene_battle_rpg2k3.cpp
index 68f163f0f..c258d9094 100644
--- a/src/scene_battle_rpg2k3.cpp
+++ b/src/scene_battle_rpg2k3.cpp
@@ -2855,21 +2855,6 @@ void Scene_Battle_Rpg2k3::RowSelected() {
 }
 
 void Scene_Battle_Rpg2k3::ActionSelectedCallback(Game_Battler* for_battler) {
-	auto single_target = for_battler->GetBattleAlgorithm()->GetOriginalSingleTarget();
-	auto group_targets = for_battler->GetBattleAlgorithm()->GetOriginalPartyTarget();
-	// Target: 0 None, 1 Single Enemy, 2 All Enemies, 3 Single Ally, 4 All Allies
-	Game_Battle::ManiacBattleHook(
-		Game_Interpreter_Battle::ManiacBattleHookType::Targetting,
-		for_battler->GetType() == Game_Battler::Type_Enemy,
-		for_battler->GetPartyIndex(),
-		for_battler->GetBattleAlgorithm()->GetActionType(),
-		for_battler->GetBattleAlgorithm()->GetActionId(),
-		single_target
-			? (single_target->GetType() != Game_Battler::Type_Enemy ? 1 : 3)
-			: (group_targets->GetRandomActiveBattler()->GetType() != Game_Battler::Type_Enemy ? 2 : 4),
-		single_target ? single_target->GetPartyIndex() : 0
-	);
-
 	for_battler->SetAtbGauge(0);
 
 	if (for_battler == active_actor) {
```

**Review by Ghabry — APPROVED** [review]

**Comment by Ghabry** [comment]
github hook test (somehow approvals are not sent?)

**Pull request merged**